federated_query/plan/physical.py

```
@@ -1188,62 +1380,101 @@ def _reduce_left_for_semi(self, right_batches: List[pa.RecordBatch]) -> None:
+    def _distinct_keys_sql(self, build_rel: str, columns: List[str]) -> str:
+        """SELECT DISTINCT of the key columns, non-NULL, capped one past the limit."""
+        quoted = []
+        for column in columns:
+            quoted.append(f'"{column}"')
+        not_null = []
+        for name in quoted:
+            not_null.append(f"{name} IS NOT NULL")
+        limit = _DYNAMIC_FILTER_MAX_KEYS + 1
+        return (
+            f"SELECT DISTINCT {', '.join(quoted)} FROM {build_rel} "
+            f"WHERE {' AND '.join(not_null)} LIMIT {limit}"
```

Ruff (0.15.20)

[error] Possible SQL injection vector through string-based query construction

(S608)